### PR TITLE
Take advantage of ccache on builds that install the dependencies

### PR DESCRIPTION
**Pull request by coryan**
Some builds need the dependencies (gRPC, protobuf, c-ares)
pre-installed.  We used to install the dependencies as part of the
Docker image creation, with this change the dependencies are
installed by the Travis script.  That allows us to use ccache
to speed up the build+install of these dependencies, just like
we do for the main code.

For the SCAN_BUILD=yes builds I could not figure out how to use
both scan-build and ccache at the same time.  Both replace CXX
and CC, so configuring that is tricky.  I was able to use
ccache for installing the dependencies (which do not require
static analysis), and then disabled ccache for running the static
analysis on our code. We can improve on this later.

For these builds we can also optimize the submodule checkout.

And finally I was spending time reading the logs to figure out what
was going on, I made them more readable to help debugging
this setup.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -35,14 +35,18 @@ matrix:
       compiler: clang
       env:
         SCAN_BUILD=yes
-        CMAKE_FLAGS=-DGOOGLE_CLOUD_CPP_GRPC_PROVIDER=package
-        DISTRO=ubuntu-with-dependencies
+        DISTRO=ubuntu
         DISTRO_VERSION=16.04
+      # Because this builds installs gRPC we can skip that submodule, that
```
Comment (paprat) [inline .travis.yml]: s/builds/build

File: .travis.yml [hunk 2]
```diff
@@ -73,13 +77,18 @@ matrix:
       compiler: clang
       env:
         TEST_INSTALL=yes
-        DISTRO=ubuntu-with-dependencies
+        DISTRO=ubuntu
         DISTRO_VERSION=17.10
+      # Because this builds installs gRPC we can skip that submodule, that
```
Comment (paprat) [inline .travis.yml]: ditto.

File: .travis.yml [hunk 3]
```diff
@@ -35,14 +35,18 @@ matrix:
...
+      # installation step.
+      git:
+        submodules: false
       install:
-        - # First build the usual Ubuntu image, then use that to build an image
-          # with gRPC already compiled and installed.
-        - DISTRO=ubuntu SCAN_BUILD=no ./ci/install-linux.sh
-        - ./ci/install-linux.sh
+        - git submodule update --init
+                    third_party/googleapis
```
Comment (mbrukman) [inline .travis.yml]: 4 space indent? This is like 8, 9 space indent? Similarly below and elsewhere.

File: ci/build-docker.sh [hunk 4]
```diff
@@ -32,31 +32,125 @@ readonly BUILD_DIR="build-output/${IMAGE}"
...
+}
+
+function install_grpc {
+  # Install gRPC. Note that we use the system's zlib and ssl libraries.
+  # For similar reasons to c-ares (see above), we need two install steps.
+  echo "${COLOR_YELLOW}Installing gRPC $(date)${COLOR_RESET}"
+  wget -q https://github.com/grpc/grpc/archive/v1.10.0.tar.gz
+  tar -xf v1.10.0.tar.gz
+  cd grpc-1.10.0
+  env CXX="${cached_cxx}" CC="${cached_cc}"  cmake \
```
Comment (mbrukman) [inline ci/build-docker.sh]: 2 spaces before `cmake` -> 1 space.

File: ci/build-docker.sh [hunk 5]
```diff
@@ -32,31 +32,125 @@ readonly BUILD_DIR="build-output/${IMAGE}"
...
+  # files for the library, so manually install it.  c-ares requires two install
+  # steps because (1) the CMake-based build does not install pkg-config files,
+  # and (2) the Makefile-based build does not install CMake config files.
+  echo "${COLOR_YELLOW}Installing c-ares $(date)${COLOR_RESET}"
+  apt-get remove -y libc-ares-dev libc-ares2
+  wget -q https://github.com/c-ares/c-ares/archive/cares-1_14_0.tar.gz
+  tar -xf cares-1_14_0.tar.gz
+  cd c-ares-cares-1_14_0
+  env CXX="${cached_cxx}" CC="${cached_cc}"  cmake \
+    -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" \
```
Comment (mbrukman) [inline ci/build-docker.sh]: 4 space indent for wrapped lines?

File: ci/build-docker.sh [hunk 6]
```diff
@@ -32,31 +32,125 @@ readonly BUILD_DIR="build-output/${IMAGE}"
...
+
+function install_grpc {
+  # Install gRPC. Note that we use the system's zlib and ssl libraries.
+  # For similar reasons to c-ares (see above), we need two install steps.
+  echo "${COLOR_YELLOW}Installing gRPC $(date)${COLOR_RESET}"
+  wget -q https://github.com/grpc/grpc/archive/v1.10.0.tar.gz
+  tar -xf v1.10.0.tar.gz
+  cd grpc-1.10.0
+  env CXX="${cached_cxx}" CC="${cached_cc}"  cmake \
+    -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" \
```
Comment (mbrukman) [inline ci/build-docker.sh]: 4 space indent for wrapped lines?

File: ci/build-docker.sh [hunk 7]
```diff
@@ -32,31 +32,125 @@ readonly BUILD_DIR="build-output/${IMAGE}"
...
+
+function install_protobuf {
+  # Install protobuf using CMake.  Some distributions include protobuf, gRPC
+  # requires 3.4.x or newer, and many of those distribution use older versions.
+  echo "${COLOR_YELLOW}Installing protobuf $(date)${COLOR_RESET}"
+  wget -q https://github.com/google/protobuf/archive/v3.5.2.tar.gz
+  tar -xf v3.5.2.tar.gz
+  cd protobuf-3.5.2/cmake
+  env CXX="${cached_cxx}" CC="${cached_cc}" cmake \
+    -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" \
```
Comment (mbrukman) [inline ci/build-docker.sh]: 4 space indent for wrapped lines?

File: ci/build-linux.sh [hunk 8]
```diff
@@ -59,8 +61,8 @@ sudo docker run \
      --env TERM="${TERM:-dumb}" \
      --user "${docker_uid}" \
      --volume "${PWD}":/v \
-     --volume "${PWD}/build-output/cache":/.cache \
-     --volume "${PWD}/build-output/ccache":/.ccache \
+     --volume "${PWD}/build-output/cache":${docker_home}/.cache \
+     --volume "${PWD}/build-output/ccache":${docker_home}/.ccache \
```
Comment (mbrukman) [inline ci/build-linux.sh]: What happens when `${docker_home}` == `""`, and this is referring to `/.cache` — should that be `$HOME/.ccache` instead?

File: .travis.yml (hunk 1 above)
Comment (coryan) [reply .travis.yml]: Fixed.

File: .travis.yml (hunk 2 above)
Comment (coryan) [reply .travis.yml]: Fixed.

File: .travis.yml (hunk 3 above)
Comment (coryan) [reply .travis.yml]: Fixed.  4 from the beginning for `git`, let me know if you prefer 4 from `- `.

File: ci/build-docker.sh (hunk 4 above)
Comment (coryan) [reply ci/build-docker.sh]: Fixed.

File: ci/build-docker.sh (hunk 5 above)
Comment (coryan) [reply ci/build-docker.sh]: Fixed.

File: ci/build-docker.sh (hunk 7 above)
Comment (coryan) [reply ci/build-docker.sh]: Fixed.

File: ci/build-linux.sh (hunk 8 above)
Comment (coryan) [reply ci/build-linux.sh]: Inside the docker image we are running with the uid of the user calling the program, but not creating an actual user inside the docker image for that uid.  Therefore there is no `$HOME` for that poor uid.

File: .travis.yml [hunk 9]
```diff
@@ -35,14 +35,18 @@ matrix:
       compiler: clang
       env:
         SCAN_BUILD=yes
-        CMAKE_FLAGS=-DGOOGLE_CLOUD_CPP_GRPC_PROVIDER=package
-        DISTRO=ubuntu-with-dependencies
+        DISTRO=ubuntu
         DISTRO_VERSION=16.04
+      # Because this build installs gRPC we can skip that submodule, that
```
Comment (mbrukman) [inline .travis.yml]: - s/gRPC we/gRPC, we/
- s/submodule, that saves/submodule, which saves us/

Similarly everywhere else this same sentence appears.

Comment (coryan) [reply .travis.yml]: Fixed in both places.

**Pull request merged**